README.md
# SiteMon
SiteMon tracks changes in site contents as set up in the ini file, using CSS selectors, Xpath, or regular expressions.
JS is supported*. Passworded pages and browser cookies are not, yet. Definitely works on [CPython 3.10.5 w64](https://www.python.org/downloads/windows/) and [PyPy 3.9 w64](https://www.pypy.org/download.html).

As output, it sends to stdout the space-separated list of "pages" that changed, as well as adds them to a report file with timestamps.

(* The script renders JS using Chromium, so it will download Chromium to `~/.pyppeteer` (or `%LOCALAPPDATA%/pyppeteer/`) on first JS request.)

## Installation
For CPython:

    python3 -m ensurepip
    python3 -m pip install --upgrade pip
    python3 -m pip install keyboard requests_html

For PyPy 3.9 (works faster):

If you're working with Windows, you'll want to install `lxml` from the corresponding wheel first.

Now, a `pp39` wheel for `lxml`... does not exist. Trying to do `pip install lxml` makes pip plop halfway. I built mine for w64, [here](https://t.me/mildlyinterestingtoo/515) (SHA1 = 7f6ca65ad4d6a268777e5c3e594afccc968447d2 :: [malware-free](https://www.virustotal.com/gui/file-analysis/OTk5Zjk1OGQ4YmE5ODM0ZjAwMDk0ZTM5MmQzYzc5YmI6MTY1NTU4NzMyMg==)). Download the file, run cmd.exe in your Download folder, then finally install the prerequisites:

    pypy3 -m ensurepip
    pypy3 -m pip install --upgrade pip
    pypy3 -m pip install lxml-4.9.0-pp39-pypy39_pp73-win_amd64.whl
    pypy3 -m pip install keyboard requests_html

If it breaks, I hope there's an official binary for py39 by then...

## Options
```
-h, --help            show this help message and exit
-c X, --config X      an ini file with the settings (default: ./sitemon.ini)
-C X, --continuity X  a JSON continuity file (default: ./sitemon.json)
-r X, --report X      the report file (default: ./sitemon.report.txt), "nul" for console-only
-f, --force           force re-check all pages
-t N, --retries N     do N retries downloading a page, -1 for infinite (default: 3)
-p, --pause           waits for a keypress before closing
-s, --sslcheck        perform SSL issue check (default: no checks)
-v, --verbose         verbose console output (multiple allowed)
-q, --quiet           no console output, overrides --verbose (multiple allowed)
```

Please read the sample config file for more instructions.


CHANGELOG.md
# SiteMon changelog

## v1.5.1

#### Minor:
  * Added the system notification (`pip install plyer` to make use of this)

## v1.5 bugfixes in ini.example, SelTest

#### Minor:
  * Fixed the [python] example in the ini.example
  * Fixed the render option and the match amount status bar info in seltest.py
  * Tested on PyPy 3.9 too, setup instructions updated
  * Readme improved
  * A useless to-do retracted for SiteMon


## v1.5

##### Major:

  * Attribute (like "href") tracking added
  * Multiple consecutive matches can now be saved to continuity (yes, you can grab page links with this as `//a[contains(@href,'youtu']` or something like that)


## v1.4

#### Major:
  * CSS selector added
  * XpathTest → SelTest
  * SelTest can test all selector types

#### Important:
  * Better handling of zero hits and missing pages
  * Default UserAgent updated
  * Engine updated to CPython 3.10.5
  * SelTest will reload the page first if you Run after changing the URL


## v1.3

#### Important:
  * Error quits also respect --pause now
  * Fixed the -vv page writeout error
  * Exception prompts do better than say "occurred at (4)" and such
  * A regex continuity now holds the first match instead of a simple boolean tracking any change, so you can see what that change was
  * Clarified the explanation for both Types in the ini example
  * -u/--update → -f/--force


## v1.2

#### Major:
  * Base script-independent Tk GUI-based Xpath testing app added\
    (* required 3rd-party library: requests_html)
    
Base script's version stays at v1.1.


## v1.1

#### Major:
  * Removed the no-JS version since you can just not request JS in the JS version for it to behave the same way
  * CLI option added to wait for a keypress before closing\
    (* requires pip install keyboard)

#### Important:
  * Retries option added for the page downloader tweaking
  * Option added for the now-optional SSL verification

#### Minor:
  * Default UserAgent updated
  * Clarified default file locations a bit for Linux-style (but still untested)


## v1.0
  * Initial release, definitely compatible with Python 3.10.4.\
  (* required 3rd-party libraries: keyboard, requests_html)

seltest.py
#!/usr/bin/env python3

# Change the following to your liking (but it'll be gone with the update~)
DefaultURL = r"https://python.org"
DefaultSel = r"/html/body"
DefaultAttr = ""
MaxResults = 20000

"""
SelTest is a Tk GUI app where you can quickly and easily test out
your element selector.

The Run button will download the page first if you haven't already,
but otherwise will simply update the selector query result.

You need to run the following prior to using the script:
  pip3 install requests_html

Conceived for personal use by Kaens Bard, 2022
Made and tested with CPython v3.10.5 win-x64. Tested on PyPy v3.9 win-x64.

Feedback: @kaens at Telegram
"""

from requests_html import HTMLSession
import tkinter as tk
from tkinter import ttk,messagebox
import warnings,sys,re
warnings.filterwarnings("ignore")
lasturl = ''
renderclicked=False

def DigitValidation(S):
    if S.isdigit():
        a = int(S)
        if a >= 0 and a <= MaxResults:
            return True
        else:
            return False
    else:
        return False

def min(a,b): return a if a<b else b

def load(url):
    global _req, page, lasturl
    warnings.filterwarnings("ignore")
    try:
        page = _req.get(url,verify=False)
    except Exception as e:
        warnings.filterwarnings("default")
        Info.set(f'Error loading page: {str(e)}')
        return(False)
    warnings.filterwarnings("default")
    Info.set("Page downloaded.")
    lasturl = url
    return(True)

def performtest(url,sel):
    global page
    global lasturl
    try:
        a = page
    except NameError:
        Info.set("No pages loaded, loading from URL field...")
        if load(url):
            Info.set("Page is loaded now, parsing...")
        else:
            return(False)

    # test passed, but what if we need to load anyway?
    if lasturl != url:
        Info.set("URL changed, reloading...")
        if load(url):
            Info.set("Page is loaded now, parsing...")
        else:
            return(False)

    Element = None
    Type = SelType.get()
    Attrib = Attr.get()
    if Type == 'xpath':
        try:
            Element = page.html.xpath(sel)
        except Exception as e:
            Info.set(f"Xpath exception: {str(e)}")
            Element = None
    elif Type == 'css':
        try:
            Element = page.html.find(sel)
        except Exception as e:
            Info.set(f"CSS Selector exception: {str(e)}")
            Element = None
    elif Type == 're':
        try:
            Element = re.findall(sel,page.text)
        except Exception as e:
            Info.set(f"RegEx exception: {str(e)}")
            Element = None
    if Element:
        lsResult.delete("1.0",tk.END)
        if IsMany.get():
            Amount = min(int(Many.get()),len(Element))
        else:
            Amount = 1
        if Type == 're':
            lsResult.insert(tk.END,'\n'.join(Element[i] for i in range(Amount)))
        elif Type in ('xpath','css'):
            if Attrib == '':
                lsResult.insert(tk.END,'\n'.join(Element[i].text for i in range(Amount)))
            else:
                lsResult.insert(tk.END,'\n'.join(Element[i].attrs[Attrib] for i in range(Amount)))
        if len(Element)>Amount:
            Info.set(f"Found {len(Element)} item(s), displaying the first {Amount}")
        else:
            Info.set(f"Found {len(Element)} item(s)")

    else:
        lsResult.delete("1.0",tk.END)
        Info.set("Can't find it")

def btLoadCallback():
    global URL
    btLoad.state = tk.DISABLED
    btRenderJS.state = tk.DISABLED
    btRun.state = tk.DISABLED
    Info.set('Attempting to load the page...')
    load(URL.get())
    btLoad.state = tk.NORMAL
    btRenderJS.state = tk.NORMAL
    btRun.state = tk.NORMAL

def btRenderJSCallback():
    global page, renderclicked
    btLoad.state = tk.DISABLED
    btRenderJS.state = tk.DISABLED
    btRun.state = tk.DISABLED
    if not renderclicked:
        if messagebox._show("Resource load warning",
          "The first time JS is used in this instance of Python, the script will download a copy of Chromium, which is around 140MB large in y2022. Continue?",
          "warning","okcancel") == 'ok':
            renderclicked = True
    if renderclicked:
        Info.set('Attempting to render the page...')
        try:
            page.html.render()
        except Exception as e:
            Info.set(f"Render exception: {str(e)}")
            del e
        else:
            Info.set("JS rendered. Run the test again.")
    btLoad.state = tk.NORMAL
    btRenderJS.state = tk.NORMAL
    btRun.state = tk.NORMAL

# when btRun pressed
def btRunCallback():
    global URL, Sel
    btLoad.state = tk.DISABLED
    btRun.state = tk.DISABLED
    performtest(URL.get(),Sel.get())
    btLoad.state = tk.NORMAL
    btRun.state = tk.NORMAL


# GUI init

rt = tk.Tk()
rt.title('Selector Test for SiteMon')
rt.minsize(400,600)
rt.geometry('400x600')
rt.columnconfigure(0,weight=1)
rt.rowconfigure(0,weight=1)

ui = ttk.Frame(rt)
ui.grid(column=0,row=0,sticky='news')
ui.rowconfigure(0,weight=0)
ui.columnconfigure(0,weight=1)
ui.rowconfigure(5,weight=1)#result

URL = tk.StringVar()
urlf = ttk.LabelFrame(ui,text='URL:')
urlf.columnconfigure(0,weight=1)
urlf.grid(column=0,row=0,sticky='new')
tbURL = ttk.Entry(urlf,textvariable=URL,exportselection=0)
URL.set(DefaultURL)
tbURL.grid(column=0,row=0,sticky='news')
tbURL.focus()

SelType = tk.StringVar()
stf = ttk.LabelFrame(ui,text='Selector type:')
stf.grid(column=0,row=1,sticky='new')
rbCSS = ttk.Radiobutton(stf,text='CSS',variable=SelType,value='css')
rbCSS.grid(column=0,row=0,sticky='new')
rbXpath = ttk.Radiobutton(stf,text='Xpath',variable=SelType,value='xpath')
rbXpath.grid(column=1,row=0,sticky='new')
rbRegex = ttk.Radiobutton(stf,text='RegEx',variable=SelType,value='re')
rbRegex.grid(column=2,row=0,sticky='new')
rbXpath.invoke()

Sel = tk.StringVar()
sf = ttk.LabelFrame(ui,text='Selector:')
sf.columnconfigure(0,weight=1)
sf.grid(column=0,row=2,sticky='new')
tbSel = ttk.Entry(sf,textvariable=Sel,exportselection=0)
Sel.set(DefaultSel)
tbSel.grid(column=0,row=0,sticky='news')

Attr = tk.StringVar()
atf = ttk.LabelFrame(ui,text='Attribute if needed (ignored for RegEx):')
atf.columnconfigure(0,weight=1)
atf.grid(column=0,row=3,sticky='new')
tbAttr = ttk.Entry(atf,textvariable=Attr,exportselection=0)
Attr.set(DefaultAttr)
tbAttr.grid(column=0,row=0,sticky='news')

btf = ttk.Frame(ui)
btf.columnconfigure(0,weight=1)
btf.columnconfigure(1,weight=1)
btf.columnconfigure(2,weight=1)
btf.grid(column=0,row=4,sticky='new')
btLoad = ttk.Button(btf,text='Load URL',command=btLoadCallback)
btLoad.grid(column=0,row=0,sticky='we')
btRenderJS = ttk.Button(btf,text='Render JS',command=btRenderJSCallback)
btRenderJS.grid(column=1,row=0,sticky='we')

btRun = ttk.Button(btf,text='Run',command=btRunCallback)
btRun.grid(column=0,row=1,sticky='we')

IsMany = tk.IntVar()
Many = tk.StringVar()
cbMany = ttk.Checkbutton(btf,text='More results:',variable=IsMany)
manyreg = ui.register(DigitValidation)
tbMany = ttk.Entry(btf, textvariable=Many, exportselection=0, validate='key',validatecommand=(manyreg,'%P'))
Many.set("5")
cbMany.grid(column=2,row=1)
tbMany.grid(column=3,row=1,sticky='e')


resf = ttk.LabelFrame(ui,text='Result')
resf.grid(column=0,row=5,sticky='news')
resf.columnconfigure(0,weight=1)
resf.rowconfigure(0,weight=1)
if sys.platform in ('linux','linux2'):
    fnt=('DejaVu Sans',10)
elif sys.platform == 'win32':
    fnt=('Calibri',10)
elif sys.platform == 'darwin':
    fnt=('SF Pro',10)
else:
    fnt=()
lsResult = tk.Text(resf,bg='#000',fg='#ccc',selectbackground='#666',selectforeground='#fff',font=fnt)
lsResult.insert(tk.END,"Here the first element matching your selector will be shown."+
    "\nWhy not play around with selectors (with https://python.org as the playground) if you're new! Here are some suggestions:"+
    "\n CSS (https://waa.ai/csssel):\na\n#content\nul.sitemap>li:last-child li\n"+
    '\n Xpath (https://devhints.io/xpath):\n//*[@id="content"]\n//div[4]\n//button[text()="Downloads"]\n//a[contains(@href,\'http:\')] , set Attribute to "href" and More results to 10\n'+
    "\n Regex (https://waa.ai/pyregex):\nmeta.*>\n")
lsResult.grid(column=0,row=0,sticky='news')

Info = tk.StringVar()
lbInfo = ttk.Label(ui,textvariable=Info,compound='text',background='#aaa')
Info.set('Input your URL and selector. Run test.')
lbInfo.grid(column=0,row=7,sticky='sew')

_req = HTMLSession()
_req.headers['User-Agent'] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36 Edg/101.0.1210.53"

try:
    from ctypes import windll
    windll.shcore.SetProcessDpiAwareness(1)

finally:
    ui.mainloop()

_req.close()

xpathtest.py
#!/usr/bin/env python3

"""
XpathTest is a Tk GUI app where you can quickly and easily test out
your element selector.

The Run button will download the page first if you haven't already,
but otherwise will simply update the Xpath query result.

You need to run the following prior to using the script:
  pip3 install requests_html

Conceived for personal use by Kaens Bard, 2022
Made and tested with CPython v3.10.4 win-x64.

Feedback: @kaens at Telegram
"""

from requests_html import HTMLSession
import tkinter as tk
from tkinter import ttk
import warnings,sys
warnings.filterwarnings("ignore")

def load(url):
    global _req
    global page
    warnings.filterwarnings("ignore")
    try:
        page = _req.get(url,verify=False)
    except Exception as e:
        warnings.filterwarnings("default")
        Info.set(f'Error loading page: {str(e)}')
        return(False)
    warnings.filterwarnings("default")
    Info.set("Page downloaded.")
    return(True)

def performtest(url,xpath):
    global page
    try:
        a = page
    except NameError:
        Info.set("No pages loaded, loading from URL field...")
        if load(url):
            Info.set("Page is loaded now, parsing...")
        else:
            return(False)
    RawElement = page.html.xpath(xpath)
    if RawElement:
        lsResult.delete("1.0",tk.END)
        lsResult.insert(tk.END,RawElement[0].text)#.split('\n'))
        Info.set(f"Found {len(RawElement)} item(s), displaying the first one")
    else:
        lsResult.delete("1.0",tk.END)
        Info.set("Can't find it")

def btLoadCallback():
    global URL
    btLoad.state = tk.DISABLED
    btRun.state = tk.DISABLED
    Info.set('Attempting to load the page...')
    load(URL.get())
    btLoad.state = tk.NORMAL
    btRun.state = tk.NORMAL

# when btRun pressed
def btRunCallback():
    global URL, Xpath
    btLoad.state = tk.DISABLED
    btRun.state = tk.DISABLED
    performtest(URL.get(),Xpath.get())
    btLoad.state = tk.NORMAL
    btRun.state = tk.NORMAL

# GUI init

rt = tk.Tk()
rt.title('Xpath Test for SiteMon')
rt.minsize(400,640)
rt.geometry('400x640')
rt.columnconfigure(0,weight=1)
rt.rowconfigure(0,weight=1)

ui = ttk.Frame(rt)
ui.grid(column=0,row=0,sticky='news')
ui.columnconfigure(0,weight=1)
ui.rowconfigure(0,weight=0)
ui.rowconfigure(1,weight=0)
ui.rowconfigure(2,weight=0)
ui.rowconfigure(3,weight=10)
ui.rowconfigure(0,weight=0)

URL = tk.StringVar()
urlf = ttk.LabelFrame(ui,text='URL:')
urlf.columnconfigure(0,weight=1)
urlf.grid(column=0,row=0,sticky='new')
tbURL = ttk.Entry(urlf,textvariable=URL,exportselection=0)
URL.set('https://python.org')
tbURL.grid(column=0,row=0,sticky='news')
tbURL.focus()

Xpath = tk.StringVar()
xpf = ttk.LabelFrame(ui,text='Xpath:')
xpf.columnconfigure(0,weight=1)
xpf.grid(column=0,row=1,sticky='new')
tbXpath = ttk.Entry(xpf,textvariable=Xpath,exportselection=0)
Xpath.set('/html/body')
tbXpath.grid(column=0,row=0,sticky='news')

btf = ttk.Frame(ui)
btf.grid(column=0,row=2,sticky='new')
btLoad = ttk.Button(btf,text='Load URL',command=btLoadCallback)
btLoad.grid(column=0,row=0)
btRun = ttk.Button(btf,text='Run',command=btRunCallback)
btRun.grid(column=0,row=1)

resf = ttk.LabelFrame(ui,text='Result')
resf.grid(column=0,row=3,sticky='news')
resf.columnconfigure(0,weight=1)
resf.rowconfigure(0,weight=1)
if sys.platform in ('linux','linux2'):
    fnt=('DejaVu Sans',10)
elif sys.platform == 'win32':
    fnt=('Calibri',10)
elif sys.platform == 'darwin':
    fnt=('SF Pro',10)
else:
    fnt=()
lsResult = tk.Text(resf,bg='#000',fg='#ccc',selectbackground='#777',selectforeground='#fff')
lsResult.grid(column=0,row=0,sticky='news')

Info = tk.StringVar()
lbInfo = ttk.Label(ui,textvariable=Info,compound='text',background='#aaa')
Info.set('Input your URL and Xpath. Load page. Run test.')
lbInfo.grid(column=0,row=4,sticky='sew')

#for c in ui.winfo_children():
#    c.grid_configure(padx=5,pady=5)

_req = HTMLSession()
_req.headers['User-Agent'] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36 Edg/101.0.1210.53"

try:
    from ctypes import windll
    windll.shcore.SetProcessDpiAwareness(1)

finally:
    ui.mainloop()

_req.close()

sitemon.py
#!/usr/bin/env python3

"""
SiteMon tracks changes in specific DOM elements on the sites
as set up in the ini file, rendering the JS in them first.*

As output, it outputs to stdout the space-separated list of "pages"
that changed, as well as adds them to a report file with timestamps.

Please run the script with "--help" and read the sample config file
for more instructions.

You can specify -v and -q several times (or like -vvqq).

(* The script renders JS using Chromium, so it will download Chromium
to ~/.pyppeteer (or %LOCALAPPDATA%/pyppeteer/) on first JS request.)

You need to run the following prior to using the script:
For CPython,
  python3 -m ensurepip
  python3 -m pip install --upgrade pip
  pip3 install keyboard requests_html

For PyPy,
  pypy3 -m ensurepip
  pypy3 -m pip install --upgrade pip
  pypy3 -m pip install keyboard requests_html

Conceived for personal use by Kaens Bard, 2022
Made and tested with CPython v3.10.5 win-x64. Tested on PyPy v3.9 win-x64.

Feedback: @kaens at Telegram
"""

SiteMonVersion = "1.5.1"

import argparse, re, os, json, struct, sys, keyboard, warnings
from datetime import datetime, timedelta
from configparser import ConfigParser
from requests_html import HTMLSession
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

def min(a,b): return a if a<b else b

def main():
    ConfFN = './sitemon.ini'
    ContFN = './sitemon.json'
    ReportFN = './sitemon.report.txt'
    UserAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36 Edg/101.0.1210.53"
    H = {} # continuity, short for 'history'
    UpdatedPages = []

    optp = argparse.ArgumentParser(prefix_chars='/-',description="Tracks changes in the specific site elements for you.\nJS works. Passworded pages and cookie logins do not.",epilog=f'SiteMon v{SiteMonVersion}')
    optp.add_argument('-c','--config', dest='ConfFile',action='store',metavar='X',help=f"an ini file with the settings (default: {ConfFN})")
    optp.add_argument('-C','--continuity', dest='ContFile',action='store',metavar='X',help=f"a JSON continuity file (default: {ContFN})")
    optp.add_argument('-r','--report', dest='ReportFile',action='store',metavar='X',help=f'the report file (default: {ReportFN}), \"nul\" for console-only')
    optp.add_argument('-f','--force', dest='F',action='store_true',help="force re-check all pages")
    optp.add_argument('-t','--retries',dest='Retries',type=int,metavar='N',help="do N retries downloading a page, -1 for infinite (default: 3)")
    optp.add_argument('-p','--pause',dest='Pause',action='store_true',help="waits for a keypress before closing")
    optp.add_argument('-s','--sslcheck',dest='VerifySSL',action='store_true',help="perform SSL issue check (default: no checks)")
    optp.add_argument('-v','--verbose', dest='V',action='count',default=0,help="verbose console output (multiple allowed)")
    optp.add_argument('-q','--quiet', dest='Q',action='count',default=0,help="no console output, overrides --verbose (multiple allowed)")
    #process arguments, optionally show help and quit:
    o = optp.parse_args()

    def AnyKey():
        if o.Pause:
            a = keyboard.read_key()

    if o.V>1 and o.Q==0:
        print(o)
    #set up the config filename
    if o.ConfFile:
        ConfFN = o.ConfFile

    if not os.path.exists(ConfFN):
        if o.Q==0:
            print('Copy the sitemon.ini.example to sitemon.ini, and make adjustments using the sample settings')
        AnyKey()
        return(1)

    # ---- load the ini file (but the commmand line takes priority) ---
    try:
        if o.Q==0:
            print('Reading configuration...')
        conf = ConfigParser()
        conf.read(ConfFN)
        if not o.ContFile:
            ContFN = conf.get('Main','ContinuityFile',fallback=ContFN)
        if not o.Retries:
            Retries = conf.get('Main','Retries',fallback=3)
        if not o.ReportFile:
            ReportFN = conf.get('Main','ReportFile',fallback=ReportFN)
        UserAgent = conf.get('Main','UserAgent',fallback=UserAgent)
        Pages = conf.get('Main','Pages',fallback='').replace(' ','').split(',')
    except Exception as e:
        if o.Q==0:
            print("Exception occurred: "+str(e))
        AnyKey()
        return(2)

    # --- override the ini file values (separate because it's outside of "try")
    if o.ContFile:
        ContFN = o.ContFile
    if o.ReportFile:
        ReportFN = o.ReportFile
    if o.Retries:
        if o.Retries < 0:
            Retries = None
        else:
            Retries = o.Retries
    if Retries == -1:
        Retries = None

    if o.V>0 and o.Q==0:
        print(f'Pages to check: {Pages}')

    if o.Q==0:
        print('Loading continuity...')
    try:
        H = json.load(open(ContFN,encoding='utf-8-sig'))
    except FileNotFoundError:
        if o.Q==0:
            print("The continuity file was not found and will be created.")
    except Exception as e:
        if o.Q==0:
            print(f"Exception loading continuity: {str(e)}")
        AnyKey()
        return(3)

    # ---- init a browser session----

    warnings.filterwarnings("ignore") # urllib3's Retry is obsolete and drops a warning
    retry_strategy = Retry(
        total=Retries, read=Retries, connect=Retries, # simplifying it all
        redirect=10,
        backoff_factor=0.2,
        status_forcelist=[413, 429, 500, 502, 503, 504],
        method_whitelist=["HEAD", "GET", "OPTIONS"]
    )
    httpadapter = HTTPAdapter(max_retries=retry_strategy)
    session = HTMLSession(mock_browser=False)
    session.mount("https://", httpadapter)
    session.mount("http://", httpadapter)
    warnings.filterwarnings("default")


    # ---- Main loop begins ---- #

    for Page in Pages:

        if o.V>0 and o.Q==0:
            print(f'Processing entry "{Page}"...')

        Updated = False

        try:
            URL = conf.get(Page,'URL')
        except:
            if o.V>0 and o.Q==0:
                print(f"{Page}: URL parameter invalid, skipping...")
            continue

        try:
            JS = conf.getboolean(Page,'JS')
        except:
            JS = False
        if o.VerifySSL:
            VerifySSL = True
        else:
            try:
                VerifySSL = conf.getboolean(Page,'VerifySSL')
            except:
                VerifySSL = False
            
        if VerifySSL:
            warnings.filterwarnings("default")
        else:
            warnings.filterwarnings("ignore")
        try:
            Referer = conf.get(Page,'Referer')
        except:
            Referer = ''
        try:
            Attribute = conf.get(Page,'Attribute')
        except:
            Attribute = ''
        try: #check for config errors, like the absence
            Amount = conf.get(Page,'Amount')
            try: #check for conversion errors in particular
                Amount = int(Amount)
            except:
                Amount = 1
            if Amount < 1: #check for sanity
                Amount = 1
        except:
            Amount = 1

        try:
            Type = conf.get(Page,'Type').lower()
        except Exception as e:
            if o.V>0 and o.Q==0:
                print(f"{Page}: {str(e)}, skipping...")
                continue
        if Type not in ('css','xpath','regex'):
            if o.V>0 and o.Q==0:
                print(f"{Page}: Type parameter invalid, skipping...")
                continue

        Search = conf.get(Page,'Search')

        if not o.F: # the following only applies if we're not force updating
            try:
                Frequency = conf.get(Page,'Frequency')
            except Exception as e:
                if o.V>0 and o.Q==0:
                    print(f"{Page}: {str(e)}, skipping...")
                    continue

            try:
                Frequency = int(Frequency)
            except:
                if o.V>0 and o.Q==0:
                    print(f"{Page}: Frequency parameter invalid, using 1 minute")
                Frequency = 1
            if Frequency <= 0:
                DTcmp = datetime.fromisoformat('9999-12-31')
            else:
                DTcmp = datetime.now() - timedelta(minutes=Frequency)

            if Page in H:
                try:
                    DT = datetime.fromisoformat(H[Page]["last_checked"])
                except:
                    DT = datetime.now()
            else:
                #a new page that must be added and not skipped
                DT = datetime.fromisoformat('1970-01-01')

            if DTcmp < DT:
                if o.V>0 and o.Q==0:
                    print(f"It's not yet the time to check {Page}, skipping...")
                continue

        # ---- if it's time to check again: ----
        try:
            if o.V>0 and o.Q==0:
                print(f'Downloading the "{Page}" page...')

            session.headers['User-Agent'] = UserAgent
            if Referer != '':
                session.headers['Referer'] = Referer
            r = session.get(URL,verify=VerifySSL)
            if JS:
                r.html.render() # executes the js
            if o.V>1 and o.Q==0:
                print(f'Saving the page to {Page}.html...')
                open(f'{Page}.html','w',encoding='utf-8').write(r.text)
        except Exception as e:
            if o.Q==0:
                print(f"Exception downloading {Page}: {str(e)}. Check your system proxy if this keeps failing. Moving on...")
            del e
            continue

        # ---- if url load succeeded: ----
        try:
            if o.V>0 and o.Q==0:
                print('Picking the requested element...')
            if o.V>1 and o.Q==0:
                print(f'The search is {Search}')
            if Type in ("css","xpath"):
                if Type == "css": # matching the CSS selector
                    RawElements = r.html.find(Search)
                elif Type == "xpath": # matching the Xpath
                    RawElements = r.html.xpath(Search)    
                if RawElements: # the further processing is identical
                    Amount = min(Amount,len(RawElements))
                    if Attribute == '':
                        Element = '\n'.join(RawElements[i].text for i in range(Amount))
                    else:
                        Element = '\n'.join(RawElements[i].attrs[Attribute] for i in range(Amount))
                else:
                    Element = ''
            elif Type == "regex": # matching the regular expression
                RawElements = re.findall(Search,r.text)
                if RawElements:
                    Element = '\n'.join(RawElements[i] for i in range(min(Amount,len(RawElements))))
                else:
                    Element = ''
            else: #nothing or some very half-implemented stuff goes here!
                Element = ''
            if o.V>0 and o.Q==0:
                if Element != None:
                    print(f'The element is {Element}')
                else:
                    print(f'Element not found')
        except Exception as e:
            if o.Q==0:
                print(f"Exception parsing {Page}: {str(e)}. Moving on...")
            del e
            continue

        # ---- if xml element pick succeeded: ----
        if Page in H:
            if H[Page]["element"] != Element:
                Updated = True
                if o.V>0 and o.Q==0:
                    print(f'Element updated since {H[Page]["element"]}')
            else:
                if o.V>0 and o.Q==0:
                    print(f'Element has not changed')
        else:
            if o.Q==0:
                print(f'Storing "{Page}" in the continuity...')
        H[Page] = {
            "last_checked": datetime.now().isoformat(),
            "element": Element
        }
        if Updated:
            UpdatedPages.append([Page,H[Page]["last_checked"]])

    # ---- Main loop ends ----

    if o.V>1 and o.Q==0:
        print(H)
    if o.Q==0:
        print(f'Saving continuity as {ContFN}...')
    try:
        json.dump(H,open(ContFN,'w',encoding='utf-8-sig'),indent=2)
    except Exception as e:
        if o.Q==0:
            print(f"Exception saving continuity: {str(e)}")
        AnyKey()
        return(4)

    if o.Q==0:
        if len(UpdatedPages)==0:
            print(f'No updates.')
        else:
            try:
                from plyer import notification
                notification.notify(title="SiteMon",
                    message="Updated pages:\n"+'\n'.join(f"{a[0]}" for a in UpdatedPages),
                    app_name="SiteMon.py")
            except:
                pass
            if ReportFN.lower() != 'nul':
                print(f'Saving the list of updated pages to {ReportFN}...')
    try:
        open(ReportFN,'a',encoding="utf-8").write('\n'.join(f"{a[0]} {a[1]}" for a in UpdatedPages))
    except Exception as e:
        if o.Q==0:
            print(f"Exception saving report file: {str(e)}")
        AnyKey()
        return(5)

    #print the updates to stdout
    if o.Q<2:
        sys.stdout.write(' '.join(a[0] for a in UpdatedPages))

    if o.V>0 and o.Q==0:
        print('All done.')

    AnyKey()


if __name__=="__main__":
    main()
